`go:embed` does not work for subdirs containing buildable subdirs

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What version of gazelle are you using?

`v0.24.0` (also tried latest commit in master)

### What version of rules_go are you using?

`v0.29.0` (would use `0.30.0` but it's broken 😄 without a devel version of Gazelle)

### What version of Bazel are you using?

`5.0.0` (also tried `4.2.1`)

### Does this issue reproduce with the latest releases of all the above?

Yeah.

### What operating system and processor architecture are you using?

Linux x86_64

### What did you do?

Try to build https://github.com/pressly/goose.

Reproduction: https://github.com/uhthomas/bazel-gazelle-issue-1192

### What did you expect to see?

Successful build.

### What did you see instead?

```
compilepkg: /home/thomas/.cache/bazel/_bazel_thomas/b57cca483c8e5e6ea6286570e14f5568/sandbox/linux-sandbox/223/execroot/__main__/external/com_github_pressly_goose_v3/goose_test.go:158:12: could not embed examples/sql-migrations/*.sql: no matching files found
FAILED: Build did NOT complete successfully
```

--- a/language/go/generate.go
+++ b/language/go/generate.go
@@ -110,7 +110,8 @@ func (gl *goLang) GenerateRules(args language.GenerateArgs) language.GenerateRes
 	var hasTestdata bool
 	for _, sub := range args.Subdirs {
 		if sub == "testdata" {
-			hasTestdata = !gl.goPkgRels[path.Join(args.Rel, "testdata")]
+			_, ok := gl.goPkgRels[path.Join(args.Rel, "testdata")]
+			hasTestdata = !ok
 			break
 		}
 	}
@@ -300,7 +301,7 @@ func (gl *goLang) GenerateRules(args language.GenerateArgs) language.GenerateRes
 	} else {
 		for _, sub := range args.Subdirs {
 			if gl.goPkgRels[path.Join(args.Rel, sub)] {
-				gl.goPkgRels[args.Rel] = true
+				gl.goPkgRels[args.Rel] = false
 				break
 			}
 		}

--- a/language/go/lang.go
+++ b/language/go/lang.go
@@ -59,7 +59,8 @@ const goName = "go"
 
 type goLang struct {
 	// goPkgDirs is a set of relative paths to directories containing buildable
-	// Go code, including in subdirectories.
+	// Go code. If the value is false, it means the directory does not contain 
+	// buildable Go code, but it has a subdir which does.
 	goPkgRels map[string]bool
 }
 

--- a/language/go/testdata/embedsrcs/m_dir/o_go/o.go
+++ b/language/go/testdata/embedsrcs/m_dir/o_go/o.go
@@ -1,0 +1,5 @@
+// m_dir/o_go/o.go should not be embedded because Gazelle will generate a build file
+// for this directory, and we can't reliably embed files across Bazel
+// packages.
+
+package o

